xla/service/gpu/gemm_rewriter.cc

```diff
@@ -858,33 +865,170 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
+  // streamline and combine the ReLU gradient computation with grad_matmul2 as
+  // an epilogue.
+  absl::Status HandleSelect(HloInstruction *instr) override {
+    HloInstruction *fwd_gemm = nullptr;
+    HloInstruction *bwd_gemm = nullptr;
+    HloInstruction *maximum = nullptr;
+    HloInstruction *compare = nullptr;
+
+    if (!Match(instr,
+               m::Select(m::Compare(&compare, CublasLtMatmulMaybeF8(&fwd_gemm),
+                                    m::Broadcast(m::ConstantScalar(0))),
+                         CublasLtMatmulMaybeF8(&bwd_gemm).WithOneUser(),
+                         m::Broadcast(m::ConstantScalar(0))))) {
+      return absl::OkStatus();
+    }
+
+    if (fwd_gemm->user_count() != 2) {
+      // Require fwd_gemm only have 2 consumers: maximum and compare.
+      return absl::OkStatus();
+    }
+
+    for (auto user : fwd_gemm->users()) {
+      if (Match(user,
+                m::MaximumAnyOrder(m::CustomCall({kCublasLtMatmulCallTarget,
+                                                  kCublasLtMatmulF8CallTarget}),
+                                   m::Broadcast(m::ConstantScalar(0))))) {
+        maximum = user;
+        break;
+      }
+    }
+    int64_t fwd_gemm_num_rows = fwd_gemm->shape().dimensions(0);
+    if (maximum == nullptr ||
+        !SupportsEpilogueFusion(fwd_gemm->shape().element_type()) ||
+        !(fwd_gemm_num_rows >= 128 && fwd_gemm_num_rows % 128 == 0)) {
+      // cublasLt requires that CUBLASLT_MATMUL_DESC_EPILOGUE_AUX_LD Must be
+      // divisible by 128 and be no less than the number of rows in the output
+      // matrix.
+      VLOG(1) << "Possible GEMM with ReLU epilogue is "
+              << "not fused into Custom Call.";
+      return absl::OkStatus();
+    }
+
+    // Rewrite Fwd Relu with ReluAux
+    TF_ASSIGN_OR_RETURN(auto gpu_config,
+                        fwd_gemm->backend_config<GpuBackendConfig>());
+    GemmBackendConfig &config = *gpu_config.mutable_gemm_backend_config();
+    if (config.beta() != 0.0) {
+      // Matrix bias add is not supported by BIAS_RELU_AUX epilogue.
+      return absl::OkStatus();
+    }
+
+    if (config.epilogue() == GemmBackendConfig::DEFAULT) {
+      config.set_epilogue(GemmBackendConfig::RELU_AUX);
+    } else if (config.epilogue() == GemmBackendConfig::BIAS) {
+      config.set_epilogue(GemmBackendConfig::BIAS_RELU_AUX);
+    } else {
+      return absl::OkStatus();
+    }
+
+    Shape mask_shape = ShapeUtil::MakeShape(
+        PrimitiveType::U8, {ShapeUtil::ElementsIn(fwd_gemm->shape()) / 8});
+    std::unique_ptr<HloInstruction> output = fwd_gemm->CloneWithNewShape(
+        ShapeUtil::MakeTupleShape({fwd_gemm->shape(), mask_shape}));
+    TF_RETURN_IF_ERROR(output->set_backend_config(gpu_config));
+    TF_RETURN_IF_ERROR(SetName(output->GetModule(), output.get()));
+    HloInstruction *tuple_output =
+        fwd_gemm->parent()->AddInstruction(std::move(output));
```

Comment:
`parent()` is necessary to ensure the correctness.

Reply:
HloInstruction::AddInstruction calls HloComputation::AddInstruction. Why do we need to call parent here?

Reply:
Ref [here](https://github.com/openxla/xla/blob/main/xla/service/gpu/gemm_rewriter.cc#L1663). Otherwise, the epilogue RELU_AUX is not seen. The `fwd_gemm` is later replaced by other operation.

Reply:
The reference link is out of date since the file has since been modfied. I'm still confused why we need to call parent() here